---
title: "Getting started with GenomicDistributions"
author: "Nathan Sheffield"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{1. Getting started with GenomicDistributions}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE}
# These settings make the vignette prettier
knitr::opts_chunk$set(results="hold", collapse=FALSE, message=FALSE)
#refreshPackage("GenomicDistributions")
#devtools::build_vignettes("code/GenomicDistributions")
#devtools::test("code/GenomicDistributions")
```

If you have a set of genomic ranges, the GenomicDistributions R package can help you with visualize some interesting properties of your regions. Currently, it can produce three kinds of plots:

- The *chromosome distribution plot*, which visualizes how your regions are distributed over chromosomes
- The *feature distance distribution plot*, which visualizes how your regions are distributed in distance to the nearest feature of interest, like Transcription Start Sites (TSSs).
- The *partition distribution plot*, which visualizes how your regions are distributed across a genomic partitioning, such as frequency of overlapping a gene body, exon, promoter, intronic, or intergenic segment.

GenomicDistributions can work with any reference genome, as long as you have some annotation data for it (like chromosome sizes, locations of genes). To make things easier for the common use cases, I've included in the package basic metadata for the most commonly used features from the reference genomes I use most (hg19, hg38, and mm10). If you need to produce similar plots with different features, partitions, or reference assemblies, that's also possible with GenomicDistributions, and not much more difficult; GenomicDistributions is very modular and will work with other bioconductor packages to process that data, but it requires one or two additional steps to curate your reference data.

In this vignette, we'll go through examples of each of the plots using my common built-in features and partitions. If you want more control, there's another advanced vignette that will introduce you how to define your own features, partitions, and chromosome sizes for custom analysis.

## Modular plotting functions in the package

Before we start, I want to explain my philosophy for designing the functions in this package. Many R plotting packages combine calculations and plotting into one giant function, but the problem with this is that if you want to take the underlying summarized data and plot it in a different way or combine it with something else, you can't. You only have access to the final plot. GenomicDistributions divides these tasks so you can use the intermediate data to design your own custom plot style if you want. Thus, each plot type has two different functions that work together to produce the plots: a *calculate* function and a *plot* function. The *calculate* functions take your GRanges object and return a table that summarizes your regions for the given analysis type. You can use these summary statistics how you like -- aggregate them across multiple region sets, insert them into other plots you have, and so forth; or, you can simply plug that result directly into the corresponding *plot* function, which returns a *ggplot2* object. Separating the calculation and plotting functions like this gives you more control over your results.

## Install

Install `GenomicDistributions` like this:

```{r, eval=FALSE}
devtools::install_github("databio/GenomicDistributions")
```

## Initialization

Start by loading up the package and getting your query set of regions as a GenomicRanges object. I've included an example bed file to demonstrate how these plots look. You can load it up like this:

```{r, echo=TRUE, results="hide", message=FALSE, warning=FALSE}
library("GenomicDistributions")
queryFile = system.file("extdata", "vistaEnhancers.bed.gz", package="GenomicDistributions")
query = rtracklayer::import(queryFile)
```

## Chromosome distribution plots

*Chromosome distribution plots* help you visualize how your regions are distributed across chromosomes. To produce these, you'll need to specify the chromosome lengths for your reference assembly. There are a few ways to do this.

For the common reference assemblies that I use (hg19, hg38, mm9, and mm10), I've included the metadata in the package. If you're working with one of these genomes, making a plot of the distribution across chromosomes takes just a couple of lines of code:

```{r chromosome-distribution}
# First, calculate the distribution:
x = calcChromBinsRef(query, "hg19")

# Then, plot the result:
plotChromBins(x)
```

What if we want to do the same thing but on 2 query sets at the same time? No problem:

```{r Chromosome distribution plots with multiple region sets}
# Let's fudge a second region set by shifting the first one over 
query2 = GenomicRanges::shift(query, 1e6)
queryList = GRangesList(vistaEnhancers=query, shifted=query2)
x2 = calcChromBinsRef(queryList, "hg19")
plotChromBins(x2)
```




These functions just do a naive binning across the genome. If you want to tweak the way the bins are handled, or use a different reference assembly, that's also possible and is only slightly more complicated. There are other functions you can use for that, which are outlined in another vignette.

## Feature distance distribution plots

*Feature distance distribution plots* will show you how your regions are distributed with respect to the nearest feature of interest. To illustrate, we'll use Transcription Start Sites (TSS) as our example feature of interest (but really, you can use any region set).

For TSS plots, since this is such a common use case, we can use a handy built-in function that does everything for us. It's just one line of code to check distances from query to your TSSs (for common genomes), and then a second line of code to plot those distances:

```{r tss-distribution, fig.cap="TSS plot. Distribution of query regions relative to TSSs", fig.small=TRUE}
# Calculate the distances:
TSSdist = calcFeatureDistRefTSS(query, "hg19")

# Then plot the result:
plotFeatureDist(TSSdist, featureName="TSS")
```

This plot uses log-scale increasing bins to show how your regions are distributed. Now, let's make a similar plot with multiple region sets input:

```{r TSS plots with multiple region sets}

TSSdist2 = calcFeatureDistRefTSS(queryList, "hg19")
plotFeatureDist(TSSdist2)

```

If you want to check distances to other features, that's no problem; `calcFeatureDistRefTSS()` is really just a wrapper for the workhorse function, `calcFeatureDist()`. To show how this works, get some features you want to check the distance to. Here, let's just shift our query set by a normally distributed random number:

```{r Build features}
featureExample = GenomicRanges::shift(query, round(rnorm(length(query), 0,1000)))
```

Now, with these features, we just use the `calcFeatureDist` function to calculate the distances. This function uses the fast rolling joins from `data.table` under the hood, so it completes very quickly. The result of this gets piped right into the plotting function as before:

```{r Calculate distances and plot, fig.cap="Feature plot. Distribution of query regions relative to arbitrary features", fig.small=TRUE}
fdd = calcFeatureDist(query, featureExample)
plotFeatureDist(fdd)
```


## Partition distribution plots

Genomic partition distribution plots show you how your regions are distributed among promoters, exons, introns, or intergenic regions. You can produce one or two-set plots like so:


```{r a partition plot}
gp = calcPartitionsRef(query, "hg19")
plotPartitions(gp)
```

Can you do 2 at a time? Yep:

```{r multiple partition plots in one}
gp2 = calcPartitionsRef(queryList, "hg19")
plotPartitions(gp2)
```

Do you want to see how your regions are distributed among other partitions, like CpG islands, enhancers, or something else? No problem: GenomicDistributions also has functions to produce priority lists from any GRanges objects.

## Cell specific open chromatin signal plots
```{r a cell specific open chromatin plot}
exampleCellMatrixFile = system.file("extdata", "example_cell_matrix.txt.gz", package="GenomicDistributions")
cellMatrix = data.table::fread(exampleCellMatrixFile)
op = calcOpenSignal(query, cellMatrix)
plotOpenSignal(op)
```

To plot multiple datasets at a time:
```{r multiple cell specific open chromatin plot}
op2 = calcOpenSignal(queryList, cellMatrix)
plotOpenSignal(op2)
```

## GC content plots 
*GC content plots* will return a probability density function of GC content percentage over the genomic ranges in the query. 

```{r GC content plots, fig.cap="GC content plot. Probability density function of GC percentage"}
# Calculate the GC content
gc1 = calcGCContentRef(query, "hg19")

# Plot the GC distribution
plotGCContent(gc1)
```


```{r multiple distributions in one}
# You can also plot the probability distribution for more than one query.
gc2 = calcGCContentRef(queryList, "hg19")
plotGCContent(gc2)
```

