---
title: "Getting started with GenomicDistributions"
author: "Nathan Sheffield"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{1. Getting started with GenomicDistributions}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE}
# These settings make the vignette prettier
knitr::opts_chunk$set(results="hold", collapse=FALSE, message=FALSE)
#refreshPackage("GenomicDistributions")
#devtools::build_vignettes("code/GenomicDistributions")
#devtools::test("code/GenomicDistributions")
```

If you have a set of genomic ranges, the GenomicDistributions R package can help you visualize properties of your region set. GenomicDistributions produces these five types of plot:

- *chromosome distribution plot* - visualizes how your regions are distributed over chromosomes
- *width distribution plot* - visualizes the distribution of range widths
- *feature distance distribution plot* - visualizes how your regions are distributed in distance to the nearest feature of interest, like Transcription Start Sites (TSSs).
- *partition distribution plot* - visualizes how your regions are distributed across a genomic partitioning, such as frequency of overlapping a gene body, exon, promoter, intronic, or intergenic segment.

GenomicDistributions can work with any reference genome, as long as you have some annotation data for it (like chromosome sizes and locations of genes). To make things easier for the common use cases, I've included in the package basic metadata for the most commonly used features from the reference genomes I use most (hg19, hg38, and mm10). If you need to produce similar plots with different features, partitions, or reference assemblies, that's also possible, and not much more difficult; GenomicDistributions is very modular and will work with other bioconductor packages to process that data, but it requires one or two additional steps to curate your reference data.

In this vignette, we'll go through examples of each of the plots using my common built-in features and partitions. If you want more control, there's another advanced vignette that will introduce you how to define your own features, partitions, and chromosome sizes for custom analysis.

## Modular plotting functions in the package

Before we start, I want to explain the design philosophy for functions in this package. Many R plotting packages combine calculations and plotting into one function. This may seem convenient, but if you want to plot the calculation results in a different way or combine them with something else, you can't because you only have access to the final plot. GenomicDistributions divides these tasks so you can use the intermediate data to design your own custom plot, or use the calculated results directly for other analysis.

In GenomicDistributions, each plot type has two functions: a *calculate* function and a *plot* function. The *calculate* functions take your GRanges object and return a table of summary results. You can use these summary statistics how you like -- aggregate them across multiple region sets, insert them into other plots you have, and so forth; or, you can simply plug that result directly into the corresponding *plot* function, which returns a *ggplot2* object. Separating the calculation and plotting functions like this gives you more control over your results.

## Install

Install `GenomicDistributions` like this:

```{r, eval=FALSE}
devtools::install_github("databio/GenomicDistributions")
```

## Initialization

Start by loading up the package and getting your query set of regions as a GenomicRanges object. I've included an example bed file to demonstrate how these plots look. You can load it up like this:

```{r, echo=TRUE, results="hide", message=FALSE, warning=FALSE}
library("GenomicDistributions")
queryFile = system.file("extdata", "vistaEnhancers.bed.gz", package="GenomicDistributions")
query = rtracklayer::import(queryFile)
```

## Chromosome distribution plots

*Chromosome distribution plots* help you visualize how your regions are distributed across chromosomes. To produce these, you'll need to specify the chromosome lengths for your reference assembly. There are a few ways to do this.

For the common reference assemblies that I use (hg19, hg38, mm9, and mm10), I've included the metadata in the package. If you're working with one of these genomes, making a plot of the distribution across chromosomes takes just a couple of lines of code:

```{r chromosome-distribution}
# First, calculate the distribution:
x = calcChromBinsRef(query, "hg19")

# Then, plot the result:
plotChromBins(x)
```

What if we want to do the same thing but on 2 query sets at the same time? No problem:

```{r Chromosome distribution plots with multiple region sets}
# Let's fudge a second region set by shifting the first one over 
query2 = GenomicRanges::shift(query, 1e6)
queryList = GRangesList(vistaEnhancers=query, shifted=query2)
x2 = calcChromBinsRef(queryList, "hg19")
plotChromBins(x2)
```




These functions just do a naive binning across the genome. If you want to tweak the way the bins are handled, or use a different reference assembly, that's also possible and is only slightly more complicated. There are other functions you can use for that, which are outlined in another vignette.

## Feature distance distribution plots

*Feature distance distribution plots* will show you how your regions are distributed with respect to the nearest feature of interest. To illustrate, we'll use Transcription Start Sites (TSS) as our example feature of interest (but really, you can use any region set).

For TSS plots, since this is such a common use case, we can use a handy built-in function that does everything for us. It's just one line of code to check distances from query to your TSSs (for common genomes), and then a second line of code to plot those distances:

```{r tss-distribution, fig.cap="TSS plot. Distribution of query regions relative to TSSs", fig.small=TRUE}
# Calculate the distances:
TSSdist = calcFeatureDistRefTSS(query, "hg19")

# Then plot the result:
plotFeatureDist(TSSdist, featureName="TSS")
```

This plot uses log-scale increasing bins to show how your regions are distributed. Now, let's make a similar plot with multiple region sets input:

```{r tss-distribution-multi, TSS plots with multiple region sets}

TSSdist2 = calcFeatureDistRefTSS(queryList, "hg19")
plotFeatureDist(TSSdist2)

```

If you want to check distances to other features, that's no problem; `calcFeatureDistRefTSS()` is really just a wrapper for the workhorse function, `calcFeatureDist()`. To show how this works, get some features you want to check the distance to. Here, let's just shift our query set by a normally distributed random number:

```{r Build features}
featureExample = GenomicRanges::shift(query, round(rnorm(length(query), 0,1000)))
```

Now, with these features, we just use the `calcFeatureDist` function to calculate the distances. This function uses the fast rolling joins from `data.table` under the hood, so it completes very quickly. The result of this gets piped right into the plotting function as before:

```{r Calculate distances and plot, fig.cap="Feature plot. Distribution of query regions relative to arbitrary features", fig.small=TRUE}
fdd = calcFeatureDist(query, featureExample)
plotFeatureDist(fdd)
```


## Partition distribution plots

Genomic partition distribution plots show you how your regions are distributed among promoters, exons, introns, or intergenic regions. You can produce one or two-set plots like so:


```{r a partition plot}
gp = calcPartitionsRef(query, "hg19")
plotPartitions(gp)
```

Can you do 2 at a time? Yep:

```{r multiple partition plots in one}
gp2 = calcPartitionsRef(queryList, "hg19")
plotPartitions(gp2)
```

Do you want to see how your regions are distributed among other partitions, like CpG islands, enhancers, or something else? No problem: GenomicDistributions also has functions to produce priority lists from any GRanges objects.

## Feature distribution plots

Feature distribution plots allow you to interpret how your reads are distributed among regions of interest. It includes two primary means of display, a *cumulative fraction of reads in features* (cFRiF) and a non-cumulative *fraction of reads in features* (FRiF) plot.

The cFRiF plot provides an information-dense look into the genomic distribution of reads relative to genomic features. It illustrates how reads are distributed among different features and, it uses a cumulative distribution to visualize how quickly the final read count is accumulated in features of a given type. To calculate the FRiF, you must provide a Granges or GRangesList objects that includes the overlap counts for each region of a particular genomic annotation. Using the built-in `calcFRiF()` function, the individual feature elements are then sorted by read count, and for each feature, we traverse the sorted list and calculate the cumulative sum of reads found in that feature divided by the total number of aligned reads. We plot the read fraction against the $log_{10}$ transformed cumulative size of all loci for each feature. This allows the identification of features that are enriched for reads with fewer total features and total genomic space.

The FRiF plot takes into account the expected fraction of each feature type and plots the $log_{10}$ of the observed/expected fractions of each genomic feature.

To visualize, let's load some example data:
```{r, echo=TRUE, results="hide", message=FALSE, warning=FALSE}
file_list = c("UTR3.bedGraph.gz", "UTR5.bedGraph.gz",
              "Enhancer.bedGraph.gz", "Exon.bedGraph.gz",
              "Intron.bedGraph.gz", "Promoter.bedGraph.gz",
              "Promoter_Flanking.bedGraph.gz")
loadFunction = function(file) {
    queryFile = system.file("extdata", file, package="GenomicDistributions")
    query = rtracklayer::import.bedGraph(queryFile)
    names(elementMetadata(query)) <- "count"
    return(query)
}
query = lapply(file_list, loadFunction)
names(query) = c("UTR3","UTR5","Enhancer","Exon","Intron","Promoter","Promoter_Flanking")
query = GRangesList(query)
```

You can then produce each plot as follows:
```{r a cFRiF plot}
cfrif = calcFRiF(query, 93805)
plotcFRiF(cfrif)
```

```{r a FRiF plot}
dists = calcExpectedDistribution(cfrif, 3099922541)
plotFRiF(dists)
```


## Cell-type-specific open chromatin signal plots
```{r a-cell-specific-open-chromatin-plot, fig.height = 6, fig.cap="Cell specific enrichment for open chromatin."}
exampleCellMatrixFile = system.file("extdata", "example_cell_matrix.txt", package="GenomicDistributions")
cellMatrix = data.table::fread(exampleCellMatrixFile)
op = calcOpenSignal(query, cellMatrix)
plotOpenSignal(op)
```

To plot multiple datasets at a time:
```{r multiple-cell-specific-open-chromatin-plot, fig.height = 7, fig.cap="Cell specific enrichment for open chromatin."}
op2 = calcOpenSignal(queryList, cellMatrix)
plotOpenSignal(op2)
```

## GC content plots 
*GC content plots* will return a probability density function of GC content percentage over the genomic ranges in the query. These plots will not be shown in the vignette because they require large BSgenome packages, which contain the full reference sequence.

```{r gc-content-plots, fig.cap="GC content plot. Probability density function of GC percentage", eval=FALSE}
# Calculate the GC content
gc1 = calcGCContentRef(query, "hg19")

# Plot the GC distribution
plotGCContent(gc1)
```


```{r gc-content-plots-multi, eval=FALSE}
# You can also plot the probability distribution for more than one query.
gc2 = calcGCContentRef(queryList, "hg19")
plotGCContent(gc2)
```

## Quantile-Trimmed Histogram of Widths
*QT histogram plots* will return a histogram of widths of the genomic ranges in the query. These plots are "Quantile-Trimmed" to preserve the distribution in the face of extreme outliars which sometimes occur in genomic ranges.

```{r QT histogram plots, fig.cap="Quantile-trimmed Histogram plot. Frequency of widths in this query"}
# Calculate the widths
qt1 = calcWidth(query)

# Plot the width distribution
plotQTHist(qt1)
```


```{r multiple qtHist plots in one}
# You can also plot the frequency distribution for more than one query.
qt2 = calcWidth(queryList)
plotQTHist(qt2)
```


```{r changing the quantiles and bins}
# Both the quantile and the bins can be manually adjusted. The quantile you indicate should be the proportion of the data that you want in each end bar of the histogram. 
plotQTHist(qt1, bins=7, quantile = .015, EndBarColor = 'black', MiddleBarColor='darkblue')
```
